### ISSUE TITLE
TypeScript 2.0.10 Compile Errors - Ignoring typescript.tsdk path for Project and Global

### ISSUE THREAD

**Issue by MarkPhillips** [Microsoft/vscode#16413]
- VSCode Version: 1.7.2
- OS Version: Windows 10

Steps to Reproduce:

1. Install TypeScript 2.0.10 globally and/or locally via npm
2. VS Code info message is "Version mismatch! global tsc (1.0.3.0) != VS Code's language service (2.0.10)..."
3. Add VS Code workspace or user setting to point to either local or global 2.0.10 tsdk folder path.
4. Error message no longer appears.
5. Build throws invalid compile errors with or without tsdk workspace or user settings.
6. OS also has Visual Studio 2013 and 2015 installed.
7. Running the command "tsc -version" in VS Code returns message "Version 1.0.3.0" 
8. Running command "tsc -version" in Visual Studio 2013 command prompt returns "Version 1.0.3.0" 
9. Visual Studio 2013 is actually using Version 1.8.5
10. Running command "tsc -version" in Visual Studio 2015 command prompt returns "Version 1.5.3" 
11. Visual Studio 2015 is actually using Version 2.0.6.0

**Comment by mjbvz** [comment]
How are you building your TypeScript project?

The `typescript.tsdk` setting only tells VSCode which version of TypeScript to use for tooling purposes and will not change which version of TypeScript is used to build your app.

To upgrade the version of TypeScript used for building, you can install TypeScript locally with:

```
npm install --save-dev typescript
```

Then use the `tsc` from there: `./node_modules/typescript/bin/tsc`


Or you can try updating the globally installed version of typescript by running:

```bash
npm install -g typescript
```

If other instances of `tsc` are installed on your PATH and `tsc --version` still returns the old version, you can also try using the full path to the new `tsc` install:

```bash
# get root path to install directory
npm root -g
# For example:  /usr/local/lib/node_modules

# Then tack on "typescript/bin/tsc"
# The full tsc path would be
/usr/local/lib/node_modules/typescript/bin/tsc
```

**Comment by MarkPhillips** [comment]
Thanks for the quick reply Matt.




I tried your suggestions for the "typescript.tsdk" setting using  both the
global and project tsc installations. Unfortunately,  VS Code complains
that path
...   node_modules\typescript\lib\tsc doesn't point to a valid tsserver
install. TypeScript language features will be disabled.

When I used the following paths, VS Code stopped complaining.  However the
build errors continue to happen:

*NPM Global Path*
"typescript.tsdk":
"/Users/MarkPhillips/AppData/Roaming/npm/node_modules/typescript/lib/"

*NPM Local Path*
"typescript.tsdk":                  "./node_modules/typescript/lib/"


From your suggestion I checked the environment variable PATH. Looks like it
is pointing to Visual Studio 2013 TSC installation at:
C:\Program Files (x86)\Microsoft SDKs\TypeScript\1.0\ .  (That explains the
tsc 1.0.3.0 version message.)

Not quite sure what to do next.  I may have to reinstall node and npm.
Previously, VSCode was working without a problem.

Regards,
Mark

**Comment by MarkPhillips** [comment]
Looks like VS Code "typescript.tsdk" settings do not override the tsc environmental path variable added by Visual Studio 2013. In other words, the path variable takes precedence over the VS Code settings.  Perhaps pointing the tsc environmental path to npm is the better choice.  Hopefully, that change plays well with Visual Studio 2013.  Visual Studio 2015 handles TypeScript differently and does not appear to cause a VS Code conflict.

**Comment by mjbvz** [comment]
@MarkPhillips Yes, sorry it should point to `typescript/lib/`, not  to `typescript/bin/tsc` like the comment originally suggested.

The `typescript.tsdk` setting only effects how VSCode picks up TypeScript for use with internal tooling. If you omit this setting, we will first check the local `node_modules` directory for a typescript install or fallback to version of TypeScript that is bundled in the VSCode distribution. Even after setting `typescript.tsdk`, you are responsible using the correct `tsc` compiler for actually compiling the code.

**Comment by MarkPhillips** [comment]
Matt,

Previously I have used gulp to compile without issue. This problem arose
while using the VS Code TSC Build Task

Unfortunately, even after pointing  typescript.tsdk  to the
typescript/lib/ directory,
the Windows10 PATH variable set by Visual Studio 2013 overrode both global
and local NPM typescript libraries and would not compile the code. Unlike
Visual Studio 2013, Visual Studio 2015 handles TypeScript differently and
makes it configurable from the project file.

Even though my local and global typescript node modules were using 2.0.10,
The VS Code Task would only build after changing the Windows 10 PATH
variable
from
     C:\Program Files (x86)\Microsoft SDKs\TypeScript\*1.0*
to
         C:\Program Files (x86)\Microsoft SDKs\TypeScript\*2.0*


So apparently the VS Code typescript.tsdk configurations are not used when
Visual Studio 2013 is installed. Ultimately, I will probably point the PATH
variable to the npm global typescript installation.

Regards,
Mark

**Comment by bdelaney** [comment]
I was tired of seeing this message too and I tried both workspace and user settings in VS Code to set the tsdk value. Nothing worked. And when I tried adding the tsdk value to User settings, I also got the warning about an invalid tsserver.

I'm on Windows with VS 2015 installed. I checked the system path variable and there is a pointer to the x86/sdks. In my case it was pointing to Typescript 2.1 which is what kept showing as my mismatch in VS code.

I found the Typescript 2.2.1 SDK download for VS 2015 (just out on 3-8-17) and installed that. VS 2015 now shows this version in it's 'about' and VS Code has stopped complaining about the global tsc mismatch.
Success!

All of the other info I came across, SO etc, never got me on the right track. This GitHub Issue is closest thing I find that identifies the problem with trying to use VS code settings to deal with this.
